gematria/datasets/process_and_filter_bbs.cc

```diff
@@ -0,0 +1,166 @@
+
+static cl::opt<std::string> InputFile(
+    "input-file",
+    cl::desc("Path to the input CSV file containing hex basic blocks"),
+    cl::init(""), cl::cat(ProcessFilterCat));
+
+static cl::opt<std::string> OutputFile(
+    "output-file",
+    cl::desc(
+        "Path to the output CSV file with processed/filtered basic blocks"),
+    cl::init(""), cl::cat(ProcessFilterCat));
+
+static cl::opt<bool> FilterMemoryAccessingBlocks(
+    "filter-memory-accessing-blocks",
+    cl::desc("Whether or not to filter out blocks that access memory"),
+    cl::init(false), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> ReportProgressEvery(
+    "report-progress-every",
+    cl::desc("The interval at which to report progress in blocks"),
+    cl::init(std::numeric_limits<unsigned>::max()), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> MaxThreadCount(
+    "thread-count",
+    cl::desc("The maximum number of threads to use to process BBs"),
+    cl::init(llvm::thread::hardware_concurrency()), cl::cat(ProcessFilterCat));
+
+static cl::opt<unsigned> BatchSize(
+    "batch-size",
+    cl::desc("The number of blocks to include in a batch to be processed by a "
+             "single thread"),
+    cl::init(1000), cl::cat(ProcessFilterCat));
+
+Expected<std::string> ProcessBasicBlock(
+    const std::string &BasicBlock,
+    const gematria::LlvmArchitectureSupport &LLVMSupport,
+    MCInstPrinter &MachineInstructionPrinter, const StringRef FileName) {
+  // TODO(boomanaiden154): Update this to use llvm::Expected once
+  // gematria::ParseHex is refactored to return llvm::Expected.
+  auto MachineCodeHex = gematria::ParseHexString(BasicBlock);
+  if (!MachineCodeHex.has_value()) {
+    return createFileError(
+        FileName,
+        make_error<StringError>(
+            llvm::Twine("Could not parse: '").concat(BasicBlock).concat("'"),
+            std::make_error_code(std::errc::invalid_argument)));
+  }
+
+  Expected<std::vector<gematria::DisassembledInstruction>>
+      DisassembledInstructionsOrErr = gematria::DisassembleAllInstructions(
+          LLVMSupport.mc_disassembler(), LLVMSupport.mc_instr_info(),
+          LLVMSupport.mc_register_info(), LLVMSupport.mc_subtarget_info(),
+          MachineInstructionPrinter, 0, *MachineCodeHex);
+
+  if (!DisassembledInstructionsOrErr)
+    return createFileError(FileName, DisassembledInstructionsOrErr.takeError());
+
+  std::string OutputBlock;
+
+  for (const gematria::DisassembledInstruction &Instruction :
+       *DisassembledInstructionsOrErr) {
+    MCInstrDesc InstDesc =
+        LLVMSupport.mc_instr_info().get(Instruction.mc_inst.getOpcode());
+    if (Instruction.mc_inst.getOpcode() == X86::SYSCALL) continue;
+    if (InstDesc.isReturn() || InstDesc.isCall() || InstDesc.isBranch())
+      continue;
+    if (FilterMemoryAccessingBlocks &&
+        (InstDesc.mayLoad() || InstDesc.mayStore()))
+      continue;
```

Comment:
IIUC, this will remove control-affecting instructions out of the block, but it preserves the rest of the instructions. I understand that/why we want to do this for branches and other control-affecting instructions, but:
1. we should check that they are at the end of the block. If they are elsewhere, we're distorting the block to something that is not realistic, and probably not correct.
2. IIUC, for blocks accessing memory, we want to drop the whole block, not just the memory-accessing instructions, which is what we do here.